src/coprocessor/codec/convert.rs

```diff
@@ -45,51 +86,157 @@ pub fn truncate_f64(mut f: f64, flen: u8, decimal: u8) -> Res<f64> {
+        Error::Eval(
+            format!("constant {} overflows {}", $val, $bound),
+            ERR_DATA_OUT_OF_RANGE,
+        )
     }};
 }
 
+/// `convert_int_to_int` converts an int value to a diferent int value.
+#[inline]
+pub fn convert_int_to_int(ctx: &mut EvalContext, val: i64, tp: FieldTypeTp) -> Result<i64> {
+    let lower_bound = integer_signed_lower_bound(tp);
+    // https://dev.mysql.com/doc/refman/8.0/en/out-of-range-and-overflow.html
+    if val < lower_bound {
+        ctx.handle_overflow(overflow!(val, lower_bound))?;
+        return Ok(lower_bound);
+    }
+    let upper_bound = integer_signed_upper_bound(tp);
+    if val > upper_bound {
+        ctx.handle_overflow(overflow!(val, upper_bound))?;
+        return Ok(upper_bound);
+    }
+    Ok(val)
+}
+
 /// `convert_uint_to_int` converts an uint value to an int value.
 #[inline]
-pub fn convert_uint_to_int(val: u64, upper_bound: i64, tp: FieldTypeTp) -> Result<i64> {
+pub fn convert_uint_to_int(ctx: &mut EvalContext, val: u64, tp: FieldTypeTp) -> Result<i64> {
+    let upper_bound = integer_signed_upper_bound(tp);
     if val > upper_bound as u64 {
-        return overflow!(val, tp);
+        ctx.handle_overflow(overflow!(val, upper_bound))?;
+        return Ok(upper_bound);
     }
     Ok(val as i64)
 }
 
 /// `convert_float_to_int` converts an f64 value to an i64 value.
 ///  Returns the overflow error if the value exceeds the boundary.
-pub fn convert_float_to_int(
-    fval: f64,
-    lower_bound: i64,
-    upper_bound: i64,
-    tp: FieldTypeTp,
-) -> Result<i64> {
+#[inline]
+pub fn convert_float_to_int(ctx: &mut EvalContext, fval: f64, tp: FieldTypeTp) -> Result<i64> {
     // TODO any performance problem to use round directly?
     let val = fval.round();
+    let lower_bound = integer_signed_lower_bound(tp);
     if val < lower_bound as f64 {
-        return overflow!(val, tp);
+        ctx.handle_overflow(overflow!(val, lower_bound))?;
+        return Ok(lower_bound);
     }
 
+    let upper_bound = integer_signed_upper_bound(tp);
     if val > upper_bound as f64 {
-        return overflow!(val, tp);
+        ctx.handle_overflow(overflow!(val, upper_bound))?;
+        return Ok(upper_bound);
     }
     Ok(val as i64)
 }
 
+/// `convert_bytes_to_int` converts a byte arrays to an i64 in best effort.
+pub fn convert_bytes_to_int(ctx: &mut EvalContext, bytes: &[u8], tp: FieldTypeTp) -> Result<i64> {
+    let s = str::from_utf8(bytes)?.trim();
+    let vs = get_valid_int_prefix(ctx, s)?;
+    let val = bytes_to_int_without_context(vs.as_bytes());
+    match val {
+        Ok(val) => convert_int_to_int(ctx, val, tp),
+        Err(_) => {
+            ctx.handle_overflow(Error::overflow("BIGINT", &vs))?;
+            let val = match vs
+                .as_bytes()
+                .iter()
+                .skip_while(|x| x.is_ascii_whitespace())
+                .next()
+            {
+                Some(&b'-') => integer_signed_lower_bound(tp),
```

Comment:
TiDB's `StrToInt` is not that complex?

Reply:
Seems we don't need to skip the whitespace. I will address it.